In [1]:
import numpy as np


In [2]:

#X represents the size of a tumor in centimeters.
X = np.array([3.78, 2.44, 2.09, 0.14, 1.72, 1.65, 4.92, 4.37, 4.96, 4.52, 3.69, 5.88]).reshape(-1,1)

#Note: X has to be reshaped into a column from a row for the LogisticRegression() function to work.
#y represents whether or not the tumor is cancerous (0 for "No", 1 for "Yes").
y = np.array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1])

In [3]:
from sklearn import linear_model

In [4]:
logr = linear_model.LogisticRegression()
logr.fit(X,y)

LogisticRegression()

predict if the tumour of size 3.46cm is cancerous

In [5]:
predicted = logr.predict(np.array([3.46]).reshape(-1,1))
print(predicted)

[0]


The value indicates that 3.46cm tumour in this set of data is not cancerous.

In Logistic Regression the coefficent is the expected change in log odds of having the outcome per unit change in X
(Shahabuddin: I dont quite understand this statement 😥 yet. will need to dig more into it)

In [6]:
log_odds = logr.coef_
odds = np.exp(log_odds)

print(odds)

[[4.03541657]]


That means size of tumour increase by 1cm the odds of it being cancerous increases by 4x

In simpler terms, from the above regression, we have the value of **w** (the optimal parameter, for this case, $w_0$ & $w_1$) available.

In [7]:
w0 = logr.intercept_
w1 = logr.coef_

print("w0: ", w0)
print("w1: ", w1)

w0:  [-4.83670029]
w1:  [[1.39510954]]


Remember for Logistic Regression, given an input X, the estimated probability that y=1 for this X is:

$h_w(x) = σ(w^Tx)$

where $σ(z) = 1/(1+e^{-z})$

Let's calculate that programatically

In [9]:
wTx = w0 + w1 * X
sigma_of_wTx = 1/(1 + np.exp(-wTx))

estimated_prob = sigma_of_wTx

print("estimated probability: \n", estimated_prob)


estimated probability: 
 [[0.60749955]
 [0.19268876]
 [0.12775886]
 [0.00955221]
 [0.08038616]
 [0.07345637]
 [0.88362743]
 [0.77901378]
 [0.88924409]
 [0.81293497]
 [0.57719129]
 [0.96664243]]


**Output interpretation**

The probability that a tumour of the size 3.78cm is cancerous is 60.75%

The probability that a tumour of the size 2.44cm is cancerous is 19.27%

The probability that a tumour of the size 2.09cm is cancerous is 12.78%

and so on.